### PR TITLE
refactor(runtime): carve rotation-proxy seams (phase 1)

### PR DESCRIPTION
**Pull request by ndycode**
## Summary

Phase 1 of carving `lib/runtime-rotation-proxy.ts` along its natural seams — audit roadmap §4.1.3 (`docs/audits/AUDIT_2026-06-10.md`, PR #522). All code moved verbatim; every previously-exported symbol re-exported, so the 3 importing call sites (`lib/index.ts`, `scripts/codex.js`, `scripts/codex-app-router.js`) and all 8 test files are untouched. **Zero behavior change**; the proxy shrinks 2,498 → 2,099 lines.

## Changes

| New module | Lines | Contents |
| --- | --- | --- |
| `lib/runtime/rotation-server-types.ts` | 69 | `RuntimeRotationProxyServer/Status/Options`, plus previously-private `RequestContext`, `ExhaustionReason`, `RuntimeProxyHttpError`, account-identity type — type-only imports, no cycle risk |
| `lib/request/stream-failover-runtime.ts` | 170 | `forwardStreamingResponse`, `readErrorBody`, `withTimeout`, client-response header filtering (`HOP_BY_HOP_HEADERS` + private/decoded-header sets) — all inputs explicit, no closure state |
| `lib/request/rate-limit-decision.ts` | 211 | The pure rate-limit/auth decision tree: retry-after parsing (header + body), quota-near-exhaustion waits, token-invalidation detection and bodies, refresh retryability, exhaustion-status normalization, pinned-unavailable error body |

**Deliberately deferred to phase 2** (entangled with closure/module state; extracting now would change call semantics): the rotation loop and `startRuntimeRotationProxy` closure (status, affinity store, stale-reload dedupe), `chooseAccount`/linear-scan fallback (mutate AccountManager cursors), `ensureFreshAccessToken` + refresh-dedupe WeakMap, and the storage-meta cache helpers.

No logging was touched (token/email-redaction rules unaffected); new modules contain zero references back to the proxy (no cycles).

## Validation

- [x] `npm run typecheck`; eslint on all 4 files `--max-warnings=0`
- [x] Proxy suites (runtime-rotation-proxy, safe-equal, issue-474 ×4): 152/154 — the 2 failures are the IPv6 `::1` bind tests, reproduced **identically** on a clean origin/main tree (environment-only)
- [x] codex-app-router + codex-bin-wrapper: 101 pass, 3 known Windows-path environment failures, identical on main
- [x] Independently re-verified: typecheck + safe-equal suite

## Risk / Rollback

Mechanical move with re-exports; revert the single commit. Phase 2 (closure-state extraction) is a separate, riskier PR by design.

https://claude.ai/code/session_01XNtnkLbBiXZxfQQYLMpucB

---
_Generated by [Claude Code](https://claude.ai/code/session_01XNtnkLbBiXZxfQQYLMpucB)_

**note:** greptile review for oc-chatgpt-multi-auth. cite files like `lib/foo.ts:123`. confirm regression tests + windows concurrency/token redaction coverage.
---

<h3>Greptile Summary</h3>

phase 1 of the `runtime-rotation-proxy.ts` carve — mechanical extraction of ~400 lines into three focused modules (`rotation-server-types.ts`, `stream-failover-runtime.ts`, `rate-limit-decision.ts`) with verbatim code moves and full re-exports at the original import path. all 3 call sites and 8 proxy test files are untouched.

- `lib/runtime/rotation-server-types.ts`: type-only extraction (server interfaces + previously-private types); no runtime code, no cycles
- `lib/request/rate-limit-decision.ts` + `lib/request/stream-failover-runtime.ts`: pure-function extraction with explicit inputs; account email/token headers remain correctly stripped from client responses via `PRIVATE_CLIENT_RESPONSE_HEADERS`
- `test/stream-failover-runtime.test.ts` pins a known pre-existing bug where a stalled upstream stream resolves as a clean end rather than triggering the `onStreamError` failover hook (documented for a follow-up PR)

<h3>Confidence Score: 5/5</h3>

safe to merge — purely mechanical moves, all previously-exported symbols re-exported, no behavior change

every function and type moved verbatim; the re-export block in runtime-rotation-proxy.ts covers the complete prior public surface; token-safety header filtering is unchanged and verified by the new test suite; the only noteworthy item is two stall tests using real 25 ms timers rather than fake timers, which is a test-convention nit and not a correctness issue

test/stream-failover-runtime.test.ts — stall tests for readErrorBody and forwardStreamingResponse rely on real wall-clock timeouts

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| lib/request/rate-limit-decision.ts | new module: verbatim extraction of rate-limit / auth-decision helpers from runtime-rotation-proxy.ts; all functions exported cleanly with no closure state or cycle risk |
| lib/request/stream-failover-runtime.ts | new module: extracted forwardStreamingResponse, readErrorBody, withTimeout, and header-filtering helpers; all inputs explicit and account email/token headers correctly stripped from client responses |
| lib/runtime/rotation-server-types.ts | new module: type-only extraction of server interfaces and previously-private types (RequestContext, ExhaustionReason, RuntimeProxyHttpError, RuntimeRotationAccountIdentity); no cycles, no runtime code |
| lib/runtime-rotation-proxy.ts | proxy shrinks ~400 lines; all previously-exported symbols re-exported via explicit re-export block; TokenResult removed from local import after moving to rate-limit-decision.ts |
| test/rate-limit-decision.test.ts | new suite: good coverage of all extracted functions including edge cases (resets_at epoch heuristics, null-index pinned-body path, date-string retry-after); uses fake timers where applicable |
| test/stream-failover-runtime.test.ts | new suite: good FakeServerResponse harness; stall tests for readErrorBody and forwardStreamingResponse use real 25 ms timers rather than vi.useFakeTimers(), contrary to project convention; withTimeout stall test correctly uses fake timers |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["lib/runtime-rotation-proxy.ts\n(main proxy — rotation loop, startRuntimeRotationProxy)"]
    B["lib/runtime/rotation-server-types.ts\n(types: Server, Status, Options,\nRequestContext, ExhaustionReason, ...)"]
    C["lib/request/rate-limit-decision.ts\n(parseRetryAfter*, isTokenInvalidation*,\nbuildTokenInvalidation*, getQuotaWait*, ...)"]
    D["lib/request/stream-failover-runtime.ts\n(forwardStreamingResponse, readErrorBody,\nwithTimeout, responseHeadersForClient)"]
    E["lib/index.ts\n(barrel — star-exports proxy)"]
    F["scripts/codex.js\nscripts/codex-app-router.js"]

    A -->|"import type"| B
    A -->|"import"| C
    A -->|"import"| D
    A -->|"re-export type (Server/Status/Options)"| B
    A -->|"re-export (buildPinnedUnavailable*, buildTokenInvalidation*)"| C
    E -->|"star re-export"| A
    F -->|"import"| E

    C -->|"import type ExhaustionReason"| B
    D -->|"import type RuntimeRotationProxyStatus"| B
```

<a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22ndycode%2Fcodex-multi-auth%22%20on%20the%20existing%20branch%20%22claude%2Faudit-15-rotation-proxy-carve%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22claude%2Faudit-15-rotation-proxy-carve%22.%0A%0AFix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Atest%2Fstream-failover-runtime.test.ts%3A1482-1488%0A**real%20timers%20in%20stall%20tests**%0A%0Atwo%20stall%20scenarios%20%E2%80%94%20the%20%60readErrorBody%60%20stall%20%28line%201482%2C%2025%20ms%20timeout%29%20and%20the%20%60forwardStreamingResponse%60%20stall%20%28line%201541%2C%2025%20ms%20timeout%29%20%E2%80%94%20rely%20on%20real%20%60setTimeout%60%20expiry.%20the%20project%20convention%20documented%20in%20%60test%2FAGENTS.md%60%20says%20stream-failover%20tests%20must%20use%20%60vi.useFakeTimers%28%29%60%20for%20deterministic%20assertions.%20the%20%60withTimeout%60%20stall%20test%20%28line%201428%29%20in%20this%20same%20file%20correctly%20does%20use%20fake%20timers.%20under%20load%2C%20a%2025%20ms%20wall-clock%20deadline%20is%20tight%20enough%20to%20produce%20intermittent%20flakes%20in%20ci.%0A%0A&repo=ndycode%2Fcodex-multi-auth&pr=532&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=3">Fix All in Codex</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 1 code review issue. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 1
test/stream-failover-runtime.test.ts:1482-1488
**real timers in stall tests**

two stall scenarios — the `readErrorBody` stall (line 1482, 25 ms timeout) and the `forwardStreamingResponse` stall (line 1541, 25 ms timeout) — rely on real `setTimeout` expiry. the project convention documented in `test/AGENTS.md` says stream-failover tests must use `vi.useFakeTimers()` for deterministic assertions. the `withTimeout` stall test (line 1428) in this same file correctly does use fake timers. under load, a 25 ms wall-clock deadline is tight enough to produce intermittent flakes in ci.

`````

</details>

<sub>Reviews (2): Last reviewed commit: ["test: add unit suites for the extracted ..."](https://github.com/ndycode/codex-multi-auth/commit/4d5dd513c2e4af6e63534532d2e0d1a7350c853c) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36667818)</sub>

**Context used:**

- Context used - test/AGENTS.md ([source](https://app.greptile.com/zeian/github/ndycode/codex-multi-auth/-/custom-context?memory=6e2636a9-e514-4bab-b249-4b4a84aa4ae3))
- Context used - lib/AGENTS.md ([source](https://app.greptile.com/zeian/github/ndycode/codex-multi-auth/-/custom-context?memory=ffbed5b7-f299-4c32-b3f9-f7c095ca8632))

### CODE REVIEW

File: lib/request/rate-limit-decision.ts [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+		normalizedPinnedIndex !== null
+			? accountSkipReasons.get(normalizedPinnedIndex) ?? null
+			: null;
+	const reasonSuffix = skipReason ? ` (${skipReason})` : "";
+	const displayIndex = (normalizedPinnedIndex ?? 0) + 1;
+	return {
+		message: `Pinned account ${displayIndex} is currently unavailable${reasonSuffix}; run \`codex-multi-auth status\` for details, or \`codex-multi-auth unpin\` to allow rotation.`,
+		code: "codex_pinned_account_unavailable",
+		pinnedAccountIndex: normalizedPinnedIndex,
+		reason: skipReason,
```
[the remark this answers is not in the document]
Comment (ndycode) [reply lib/request/rate-limit-decision.ts:203]: Valid finding, but it's pre-existing behavior this PR moved verbatim (the same `(normalizedPinnedIndex ?? 0) + 1` lives on `main` inside `runtime-rotation-proxy.ts`), so the fix lands in the stacked PR #546 to keep this one zero-behavior-change: the null-index desync path now says "The pinned account is currently unavailable…" and the previously-pinned contradictory expectation in `test/issue-474-pin-honored.test.ts` is updated there.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/request/stream-failover-runtime.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	"trailer",
+	"transfer-encoding",
+	"upgrade",
+]);
+const PRIVATE_CLIENT_RESPONSE_HEADERS = new Set([
+	"x-codex-multi-auth-account-index",
+	"x-codex-multi-auth-account-label",
+	"x-codex-multi-auth-account-email",
+	"x-codex-multi-auth-account-id",
+]);
```
[the remark this answers is not in the document]
Comment (ndycode) [reply lib/request/stream-failover-runtime.ts:21]: Agreed — done in the stacked PR #546 (this PR is a verbatim move, so the hardening goes there): the exact-name set is replaced by a `x-codex-multi-auth-account-` prefix block, with a regression test covering an unknown future header under the prefix (including mixed case).

---
_Generated by [Claude Code](https://claude.ai/code)_
